includes/server/class-llms-rest-memberships-controller.php

```diff
@@ -0,0 +1,590 @@
+
+		// Sales page id.
+		if ( 'page' === $data['sales_page_type'] || 'edit' === $context ) {
+			$data['sales_page_page_id'] = $membership->get( 'sales_page_content_page_id' );
+		}
+
+		// Sales page url.
+		if ( 'custom' === $data['sales_page_type'] || 'edit' === $context ) {
+			$data['sales_page_url'] = $membership->get( 'sales_page_content_url' );
+		}
+
+		return $data;
+	}
+
+	/**
+	 * Register routes.
+	 *
+	 * @since [version]
+	 *
+	 * @return void
+	 */
+	public function register_routes() {
+		parent::register_routes();
+
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/(?P<id>[\d]+)/enrollments',
+			array(
+				'args'   => array(
+					'id' => array(
+						'description' => __( 'Unique Membership Identifier. The WordPress Post ID', 'lifterlms' ),
+						'type'        => 'integer',
+					),
+				),
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this->enrollments_controller, 'get_items' ),
+					'permission_callback' => array( $this->enrollments_controller, 'get_items_permissions_check' ),
+					'args'                => $this->enrollments_controller->get_collection_params(),
+				),
+				'schema' => array( $this->enrollments_controller, 'get_public_item_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Updates a single llms membership.
+	 *
+	 * @since [version]
+	 *
+	 * @param LLMS_Membership $membership    LLMS_Membership instance.
+	 * @param WP_REST_Request $request       Full details about the request.
+	 * @param array           $schema        The item schema.
+	 * @param array           $prepared_item Array.
+	 * @param bool            $creating      Optional. Whether we're in creation or update phase. Default true (create).
+	 * @return bool|WP_Error True on success or false if nothing to update, WP_Error object if something went wrong during the update.
+	 */
+	protected function update_additional_object_fields( $membership, $request, $schema, $prepared_item, $creating = true ) {
+		$error = new WP_Error();
+
+		// Auto enroll.
+		if ( ! empty( $schema['properties']['auto_enroll'] ) && isset( $request['auto_enroll'] ) ) {
+			$membership->add_auto_enroll_courses( $request['auto_enroll'], true );
+		}
+
+		// Catalog visibility.
+		if ( ! empty( $schema['properties']['catalog_visibility'] ) && isset( $request['catalog_visibility'] ) ) {
+			$membership->get_product()->set_catalog_visibility( $request['catalog_visibility'] );
+		}
+
+		// Instructors.
+		if ( ! empty( $schema['properties']['instructors'] ) && isset( $request['instructors'] ) ) {
+
+			$instructors = array();
+
+			foreach ( $request['instructors'] as $instructor_id ) {
+				$user_data = get_userdata( $instructor_id );
+				if ( ! empty( $user_data ) ) {
+					$instructors[] = array(
+						'id'   => $instructor_id,
+						'name' => $user_data->display_name,
+					);
+				}
+			}
+
+			$membership->set_instructors( $instructors );
+		}
+
+		$to_set = array();
+
```

Comment:
Mmm, this `$to_set` is always empty here...
There's a whole part missing here in my opinion, see:
https://github.com/eri-trabiccolo/lifterlms-rest/blob/memberships/includes/server/class-llms-rest-memberships-controller.php#L473-L515

Let's first talk about the `restriction_notice` property.
The reason why I implemented that "placeholder" and placeholder substitution thing is that I wanted to have consistency between a membership creation with the rest-api and a membership creation via the WP backend.
In WP, when you create a membership, the restriction notice is always set as membership meta. So should do the rest-api creation IMHO. What do you think?
That's basically the reason behind my choice of handling the "default" here and not in  `prepare_object_for_response()`.

The other missing part regards to the `restriction_add_notice` membership object property. While this property is not in the rest-api spec, we still need to set it in a way. Please see:
https://github.com/eri-trabiccolo/lifterlms-rest/blob/memberships/includes/server/class-llms-rest-memberships-controller.php#L486-L489
and
https://github.com/eri-trabiccolo/lifterlms-rest/blob/memberships/includes/server/class-llms-rest-memberships-controller.php#L506-L515

Reply:
Thank you for catching this. Your knowledge and experience in LifterLMS is great!

Should the [`{{membership_id}}` placeholder be replaced](https://github.com/eri-trabiccolo/lifterlms-rest/blob/memberships/includes/server/class-llms-rest-memberships-controller.php#L503) anytime the given restriction message is give, on both creating and updating?

I've added setting `restriction_notice` to the same default value that the admin edit membership page uses when creating a membership without giving `restriction_message`.

Reply:
Thanks! I think you know a lot of llms too, it's just that I had to study a little more the llms posts to build the first api resources :) 

I've invented and the "placeholder" thing for the only purpose of setting the default value and only on creation. 
On updating you know the membership/course id and you can (and have) to provide it in plain if you want to use it: that was my opinion.
Basically the placeholder replacement thing is a response to a need, not a new feature the api consumer can use :).

Reply:
Makes sense. I agree.